keras_nlp/tokenizers/sentence_piece_tokenizer_trainer.py

```diff
@@ -0,0 +1,142 @@
+            lowercased before tokenization.
+
+    Returns:
+        Returns a string proto_output_file to a SentencePiece proto file or a
+        bytes object with a serialized SentencePiece proto
+
+    Examples:
+
+    Basic Usage (from Dataset).
+    >>> inputs = tf.data.Dataset.from_tensor_slices(["Drifting Along"])
+    >>> proto = compute_sentencepiece_vocabulary(inputs, vocabulary_size = 15)
+    >>> tokenizer = keras_nlp.tokenizers.SentencePieceTokenizer(proto=proto)
+    >>> outputs = inputs.map(tokenizer)
+    >>> for output in outputs:
+    ...   print(output)
+    tf.Tensor([ 4  8 12  5  9 14  5  6 13  4  7 10 11  6 13],
+    shape=(15,), dtype=int32)
+
+    # Basic Usage (from filenames).
+    >>> with open("test.txt", "w+") as f: f.write("Drifting Along\n")
+    >>> inputs = ["test.txt"]
+    >>> proto = compute_sentencepiece_vocabulary(inputs, 15)
+
+    # Basic Usage (with multiple files).
+    >>> with open("test1.txt", "w+") as f: f.write("Drifting Along\n")
+    >>> with open("test2.txt", "w+") as f: f.write("Woah look there\n")
+    >>> inputs = ["test1.txt", "test2.txt"]
+    >>> proto = compute_sentencepiece_vocabulary(inputs, 20)
+
+    # Basic Usage (with proto_output_file)
+    >>> inputs = tf.data.Dataset.from_tensor_slices(["Drifting Along"])
+    >>> proto = compute_sentencepiece_vocabulary(inputs, vocabulary_size = 15,
+    ...     proto_output_file = "model.spm")
+    >>> tokenizer = keras_nlp.tokenizers.SentencePieceTokenizer(proto="model.spm")
+    >>> ds = tf.data.Dataset.from_tensor_slices(["the quick brown fox."])
+    >>> ds = ds.map(tokenizer)
+    >>> for output in ds:
+    ...   print(output)
+    tf.Tensor([ 4 14  1  4  1  5  1  4  1 12 11  1  6  4  9 11  1],
+    shape=(17,), dtype=int32)
+
+    # Usage with lowercase
+    >>> inputs = tf.data.Dataset.from_tensor_slices(["Drifting Along"])
+    >>> proto = compute_sentencepiece_vocabulary(inputs, vocabulary_size = 15, lowercase =True)
+    >>> tokenizer = keras_nlp.tokenizers.SentencePieceTokenizer(proto=proto)
+    >>> outputs = inputs.map(tokenizer)
+    >>> for output in outputs:
+    ...   print(output)
+    tf.Tensor([ 4  8 12  5  9 14  5  6 13  4  7 10 11  6 13],
+    shape=(15,), dtype=int32)
+    """
+
+    if not isinstance(data, (list, tf.data.Dataset)):
+        raise ValueError(
+            "The `data` argument must be either `tf.data.Dataset` or `list`. "
+            f"Received: {type(data)}."
+        )
+
+    if model_type not in ["unigram", "bpe", "word", "char"]:
+        raise ValueError(
+            "The `model_type` argument must be one of `unigram`, `bpe`, `word`"
+            "or `char``."
+            f"Received: {model_type}."
+        )
+
+    model_writer = (
+        open(proto_output_file, "wb") if proto_output_file else io.BytesIO()
+    )
+    if isinstance(data, tf.data.Dataset):
+        spm.SentencePieceTrainer.train(
+            sentence_iterator=data.as_numpy_iterator(),
```

Comment:
Curious - must it be a numpy iterator? can it just be `iter(data)`?

Reply:
@chenmoneygithub It causes an error as it feeds tensors instead of strings:
<img width="324" alt="image" src="https://user-images.githubusercontent.com/72096386/182637812-4ebf06dc-5f04-4892-ad80-45db8f6c1057.png">
The error raised is: 
```
[/usr/local/lib/python3.7/dist-packages/sentencepiece/__init__.py](https://localhost:8080/#) in _TrainFromMap4(args, iter)
    399     @staticmethod
    400     def _TrainFromMap4(args, iter):
--> 401         return _sentencepiece.SentencePieceTrainer__TrainFromMap4(args, iter)
    402 
    403     @staticmethod

RuntimeError: Internal: Not a string.
```
I reckon this is because it isn't designed to take tensor inputs